### PR TITLE
Fixes REPL initialization and SourcePaths handling

### PR DESCRIPTION
**Pull request by tmat**
Fixes bug https://github.com/dotnet/roslyn/issues/5438.
Fixes #load completion to consider source search paths.
Fixes .rsp file processing to initialize search paths and imports correctly.

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/interactive

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 1]
```diff
@@ -197,40 +204,38 @@ private void ProcessStarting(bool initialize)
...
+            SourceSearchPaths = ImmutableArray<string>.Empty;
 
             if (initialize && File.Exists(_responseFilePath))
             {
                 // The base directory for relative paths is the directory that contains the .rsp file.
                 // Note that .rsp files included by this .rsp file will share the base directory (Dev10 behavior of csc/vbc).
-                var rspArguments = this.CommandLineParser.Parse(new[] { "@" + _responseFilePath }, Path.GetDirectoryName(_responseFilePath), RuntimeEnvironment.GetRuntimeDirectory(), null /* TODO: pass a valid value*/);
-                ReferenceSearchPaths = ReferenceSearchPaths.AddRange(rspArguments.ReferencePaths);
+                var rspDirectory = Path.GetDirectoryName(_responseFilePath);
+                var args = this.CommandLineParser.Parse(new[] { "@" + _responseFilePath }, rspDirectory, RuntimeEnvironment.GetRuntimeDirectory(), null);
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: named args would be helpful here

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 2]
```diff
@@ -446,7 +455,7 @@ public virtual bool CanExecuteCode(string text)
         public Task<ExecutionResult> ResetAsync(bool initialize = true)
         {
             GetInteractiveWindow().AddInput(_interactiveCommands.CommandPrefix + "reset");
-            GetInteractiveWindow().WriteLine("Resetting execution engine.");
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: Woops--how'd I miss that one...? :frowning:

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 3]
```diff
@@ -197,40 +204,38 @@ private void ProcessStarting(bool initialize)
...
-                ReferenceSearchPaths = ReferenceSearchPaths.AddRange(rspArguments.ReferencePaths);
+                var rspDirectory = Path.GetDirectoryName(_responseFilePath);
+                var args = this.CommandLineParser.Parse(new[] { "@" + _responseFilePath }, rspDirectory, RuntimeEnvironment.GetRuntimeDirectory(), null);
 
-                // the base directory for references specified in the .rsp file is the .rsp file directory:
-                var rspMetadataReferenceResolver = CreateMetadataReferenceResolver(metadataService, ReferenceSearchPaths, rspArguments.BaseDirectory);
+                if (args.Errors.Length == 0)
+                {
+                    var metadataResolver = CreateMetadataReferenceResolver(metadataService, args.ReferencePaths, rspDirectory);
+                    var sourceResolver = CreateSourceReferenceResolver(args.SourcePaths, rspDirectory);
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: where is ```sourceResolver``` used?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 4]
```diff
@@ -90,11 +91,17 @@ public abstract class InteractiveEvaluator : IInteractiveEvaluator, ICurrentWork
             _commandsFactory = commandsFactory;
             _commands = commands;
 
-            var hostPath = interactiveHostPath;
-            _interactiveHost = new InteractiveHost(replType, hostPath, initialWorkingDirectory);
-            _interactiveHost.ProcessStarting += ProcessStarting;
-
+            // The following settings will apply when the REPL starts without .rsp file.
+            // They are discarded once the REPL is reset.
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: "They are discarded by the default reset (which uses the .rsp file)."? (or something to that effect...it's not completely clear why these would be discarded)

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 5]
```diff
@@ -90,11 +91,17 @@ public abstract class InteractiveEvaluator : IInteractiveEvaluator, ICurrentWork
...
+            ReferenceSearchPaths = ImmutableArray<string>.Empty;
```
Comment (ManishJayaswal) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: What is the expected behavior here? repl will start without .rsp file when noconfig option is used. Then if a reset is done these would be discarded? If, yes then why? What would be used instead?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs (hunk 4 above)
Comment (tmat) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: They will be overridden by the reset logic.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 6]
```diff
@@ -90,11 +91,17 @@ public abstract class InteractiveEvaluator : IInteractiveEvaluator, ICurrentWork
...
+            SourceSearchPaths = ImmutableArray<string>.Empty;
             WorkingDirectory = initialWorkingDirectory;
+            var metadataService = _workspace.CurrentSolution.Services.MetadataService;
+            _metadataReferenceResolver = CreateMetadataReferenceResolver(metadataService, ReferenceSearchPaths, _initialWorkingDirectory);
+            _sourceReferenceResolver = CreateSourceReferenceResolver(SourceSearchPaths, _initialWorkingDirectory);
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: So these weren't getting initialized *at all* before?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs (hunk 5 above)
Comment (tmat) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: These are just default values for an edge case scenario, when the .rsp file is actually deleted. Whenever the REPL is restarted via #reset command a different code path sets up the context.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs (hunk 3 above)
Comment (tmat) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: Unused. Removed.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs (hunk 6 above)
Comment (tmat) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: I don't think they were.

**Comment by KevinH-MS** [comment]
:+1:

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs (hunk 6 above)
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: Eek...yeah, definitely seems like they should have been initialized here.

**Pull request merged**